modules/raft/src/main/java/org/apache/ignite/internal/raft/RetryContext.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.raft;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import org.apache.ignite.internal.network.NetworkMessage;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Represents a context containing data for {@code RaftGroupServiceImpl#sendWithRetry} methods.
+ *
+ * <p>Not thread-safe.
+ */
+class RetryContext {
```

Comment:
It's not thread safe because it's expected to be used in ScheduledExecutorService only? And executor is expected to be single threaded, right? Even if it's true - it's rather fragile. I'd rather add several volatiles to traceId etc.

Reply:
How do you guarantee e.g. contextTraceId visibility? handle<> might be called within different threads, thus traceId initialisation might not be visible by another thread. Where am I wrong?

Reply:
As you wrote 

> Memory consistency effects: Actions in a thread prior to submitting a Runnable object to an Executor happen-before its execution begins, perhaps in another thread

so, there's probably no race.